src/q2g-ext-bookmarkDirective.ts

```diff
@@ -0,0 +1,590 @@
+
+//#region IMPORT
+import { Logging } from "./lib/daVinci.js/src/utils/logger";
+import { ListViewDirectiveFactory, IDataModelItem } from "./lib/daVinci.js/src/directives/listview";
+import { IdentifierDirectiveFactory } from "./lib/daVinci.js/src/directives/identifier";
+import { Q2gListAdapter, Q2gListObject, Q2gIndObject } from "./lib/daVinci.js/src/utils/object";
+import { ScrollBarDirectiveFactory } from "./lib/daVinci.js/src/directives/scrollBar";
+import { ShortCutDirectiveFactory } from "./lib/daVinci.js/src/directives/shortcut";
+import { ExtensionHeaderDirectiveFactory } from "./lib/daVinci.js/src/directives/extensionHeader";
+
+import { calcNumbreOfVisRows,
+         templateReplacer,
+         checkDirectiveIsRegistrated,
+         AssistHyperCubeBookmarks,
+         IMenuElement,
+         StateMachineInput,
+         IStateMachineState}
+    from "./lib/daVinci.js/src/utils/utils";
+import * as template from "text!./q2g-ext-bookmarkDirective.html";
+//#endregion
+
+//#region INTERFACES
+interface IShortcutProperties {
+    shortcutFocusBookmarkList: string;
+    shortcutFocusSearchField: string;
+    shortcutRemoveBookmark: string;
+    shortcutAddBookmark: string;
+}
+//#endregion
+
+let logger = new Logging.Logger("q2g-ext-selectorDrective");
+
+class BookmarkController implements ng.IController {
+
+    $onInit(): void {
+        logger.debug("initialisation from BookmarkController");
+    }
+
+    //#region Variables
+    actionDelay: number = 0;
+    bookmarkList: Q2gListAdapter;
+    editMode: boolean = false;
+    element: JQuery;
+    focusedPosition: number = 1;
+    inputBarType: string = "search";
+    menuList: Array<IMenuElement>;
+    properties: IShortcutProperties = {
+        shortcutFocusBookmarkList: " ",
+        shortcutFocusSearchField: " ",
+        shortcutRemoveBookmark: " ",
+        shortcutAddBookmark: " "
+    };
+    showButtons: boolean = false;
+    showFocused: boolean = true;
+    showSearchField: boolean = false;
+    timeout: ng.ITimeoutService;
+    titleDimension: string = "Bookmarks";
+    selectBookmarkToggle: boolean = true;
+    inputStates: StateMachineInput = new StateMachineInput();
+    inputBarFocus: boolean = false;
+    //#endregion
+
+    //#region elementHeight
+    private _elementHeight: number = 0;
+    get elementHeight(): number {
+        return this._elementHeight;
+    }
+    set elementHeight(value: number) {
+        if (this.elementHeight !== value) {
+            try {
+                this._elementHeight = value;
+                if (this.bookmarkList) {
```

Comment:
the first if in the setter is, to check if the element height realy changed, so that the inner code will not be called reduntend. The inner if is important, to check if the bookmark list is initialized. It is not possible to include both if statements in one, cause i need to seht the element height. But i changed the code to:

   set elementHeight(value: number) {
        if (this.elementHeight !== value) {
            try {
                this._elementHeight = value;
                if (this.bookmarkList && this.bookmarkList.obj) {
                    this.bookmarkList.obj.emit("changed", calcNumbreOfVisRows(this.elementHeight));
                }
            } catch (err) {
                this.logger.error("error in setter of elementHeight", err);
            }
        }
    }